finality-aleph/src/data_io/status_provider.rs

```diff
@@ -0,0 +1,397 @@
+
+    fn on_block_finalized(&mut self, block: BlockHashNum<B>);
+
+    fn is_block_imported(&mut self, block: &BlockHashNum<B>) -> bool;
+
+    fn get_finalized_at(&mut self, number: NumberFor<B>) -> Result<BlockHashNum<B>, ()>;
+
+    fn get_parent_hash(&mut self, block: &BlockHashNum<B>) -> Result<B::Hash, ()>;
+
+    // Default implementation is provided but can be overriden to enable caching.
+    fn get_status(
+        &mut self,
+        proposal: &AlephProposal<B>,
+        old_status: Option<&ProposalStatus>,
+    ) -> ProposalStatus {
+        self.get_status_inner(proposal, old_status)
+    }
+
+    fn is_hopeless_fork(&mut self, proposal: &AlephProposal<B>) -> bool {
+        let bottom_num = proposal.number_bottom_block();
+        for i in 0..proposal.branch.len() {
+            if let Ok(finalized_block) =
+                self.get_finalized_at(bottom_num + <NumberFor<B>>::saturated_from(i))
+            {
+                if finalized_block.hash != proposal.branch[i] {
+                    return true;
+                } else {
+                    // We don't know the finalized block at this height
+                    break;
+                }
+            }
+        }
+        false
+    }
+
+    fn is_ancestor_finalized(&mut self, proposal: &AlephProposal<B>) -> bool {
+        let bottom = proposal.bottom_block();
+        let parent_hash = if let Ok(hash) = self.get_parent_hash(&bottom) {
+            hash
+        } else {
+            return false;
+        };
+        let finalized = if let Ok(hash) = self.get_finalized_at(proposal.number_below_branch()) {
+            hash
+        } else {
+            return false;
+        };
+        parent_hash == finalized.hash
+    }
+
+    fn is_branch_correct(&mut self, proposal: &AlephProposal<B>) -> bool {
+        let bottom_num = proposal.number_bottom_block();
+        for i in 1..proposal.branch.len() {
+            let curr_block = (
+                proposal.branch[i],
+                bottom_num + <NumberFor<B>>::saturated_from(i),
+            )
+                .into();
+            match self.get_parent_hash(&curr_block) {
+                Ok(parent_hash) => {
+                    if parent_hash != proposal.branch[i - 1] {
+                        return false;
+                    }
+                }
+                Err(()) => {
+                    warn!(target: "afa", "A proposal whose top block is available but the whole branch is not {:?}", proposal);
+                    return false;
+                }
+            }
+        }
+        true
+    }
+
+    fn get_status_inner(
+        &mut self,
+        proposal: &AlephProposal<B>,
+        old_status: Option<&ProposalStatus>,
+    ) -> ProposalStatus {
+        use crate::data_io::proposal::AvailableKind::*;
+        use crate::data_io::proposal::ProposalStatus::*;
+
+        if self.is_hopeless_fork(proposal) {
+            return Available(HopelessFork);
```

Comment:
I do not see the reason why the (whole) branch is available here

Reply:
Perhaps the status of `HopelessFork` should be separated from being `Available`

Reply:
There is a definition of `available proposal` somewhere in the `data_store.rs` file as a comment, and according to the definition `HopelessFork` certainly counts as `available`. The point here is that "available" means that we should let it through, not that the proposal is correct and all its blocks are available.